### ISSUE TITLE
Confusing assertion error when building docs without git installed

### ISSUE THREAD

**Issue by myfreeweb** [crystal-lang/crystal#9789]
Building 0.35.1 in a clean environment (FreeBSD packaging) would fail with a confusing

```
bin/crystal docs src/docs_main.cr  --project-name=Crystal --project-version=0.35.1 --source-refname=
bin/crystal: crystal: not found
Using compiled compiler at .build/crystal
Unhandled exception: Crystal::Compiler#program cannot be nil (NilAssertionError)
  from __crystal_realloc
  from __crystal_sigfault_handler
  from __crystal_sigfault_handler
  from __crystal_main
  from main
  from _start
```

turns out, right after it tries to execute git:

```
12598: execve("/usr/local/sbin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: execve("/usr/local/bin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: execve("/usr/sbin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: execve("/usr/bin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: execve("/sbin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: execve("/bin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: write(15,"\^A",1)                         = 1 (0x1)
 7355: kevent(15,{ 14,EVFILT_READ,EV_ADD,0,0,0x30303 8,EVFILT_READ,EV_ADD,0,0,0x30303 11,EVFILT_READ,EV_ADD,0,0,0x30303 },3,{ 14,EVFILT_READ,0x0,0,0x1,0x30303 },64,{ 4.999951000 }) = 1 (0x1)
12598: write(15,"\^B\0\0\0",4)                   = 4 (0x4)
 7355: read(14,"\^A\^B\0\0\0",8192)              = 5 (0x5)
 7355: mmap(0x0,28672,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34508636160 (0x808e00000)
 7355: fstatat(AT_FDCWD,"/usr/share/nls/C/libc.cat",0x7fffffffbcb0,0x0) ERR#2 'No such file or directory'
 7355: fstatat(AT_FDCWD,"/usr/share/nls/libc/C",0x7fffffffbcb0,0x0) ERR#2 'No such file or directory'
 7355: fstatat(AT_FDCWD,"/usr/local/share/nls/C/libc.cat",0x7fffffffbcb0,0x0) ERR#2 'No such file or directory'
 7355: fstatat(AT_FDCWD,"/usr/local/share/nls/libc/C",0x7fffffffbcb0,0x0) ERR#2 'No such file or directory'
 7355: close(14)                                 = 0 (0x0)
Unhandled exception: Crystal::Compiler#program cannot be nil (NilAssertionError)
```

and installing git fixes it.

Would be nice to make git not required (probably there's already a way to provide the git rev or whatever it wants?) but at least please make a **good error message**!

**Comment by bcardiff** [comment]
You found the culprint of the build process on freebsd. That is great.

I agree that either git should be optional or a better error message should be introduced.

You are able to unlock the 0.35.1 package on freebsd then, right?

---

For people that do not know @myfreeweb is the maintainer of the crystal and shards port on freebsd.

* https://svnweb.freebsd.org/ports/head/lang/crystal/
* https://svnweb.freebsd.org/ports/head/devel/shards/

**Comment by straight-shoota** [comment]
Hm, the strict dependency on git should have been removed in 0.35.0 with #9119 and #8792 (maybe also others I don't remember). The doc generator should definitely work without git installed. I know I checked that and it worked.

However, this issue definitely reproduces:

```console
$ sudo docker run --rm -it crystallang/crystal:0.35.1 bash
root@2f26525752d5:/# apt remove -yq git
Reading package lists...
Building dependency tree...
Reading state information...
The following packages were automatically installed and are no longer required:
  git-man krb5-locales less libbsd0 libcurl3-gnutls libedit2 liberror-perl libgssapi-krb5-2 libk5crypto3 libkeyutils1 libkrb5-3 libkrb5support0 libnghttp2-14 libpsl5 librtmp1 libssl1.0.0 libx11-6 libx11-data libxau6 libxcb1 libxdmcp6
  libxext6 libxmuu1 multiarch-support openssh-client publicsuffix xauth
Use 'apt autoremove' to remove them.
The following packages will be REMOVED:
  git
0 upgraded, 0 newly installed, 1 to remove and 0 not upgraded.
After this operation, 32.3 MB disk space will be freed.
(Reading database ... 19825 files and directories currently installed.)
Removing git (1:2.17.1-1ubuntu0.7) ...
root@2f26525752d5:/# crystal docs
Unhandled exception: Crystal::Compiler#program cannot be nil (NilAssertionError)
  from crystal/src/compiler/crystal/compiler.cr:157:5 in 'report_warnings'
  from crystal/src/compiler/crystal/command.cr:142:5 in 'run'
  from crystal/src/compiler/crystal.cr:11:1 in '__crystal_main'
  from crystal/src/crystal/main.cr:105:5 in 'main'
  from src/env/__libc_start_main.c:94:2 in 'libc_start_main_stage2'
```

**Comment by straight-shoota** [comment]
So it seems `Crystal::Compiler#program cannot be nil` is only a secondary error. It occurs because there's an error in the docs generator and the error handling causes `report_warnings` to be called. `report_warnings` expects `@compiler` to have the `program` property. But this property has not been set because of the original error. Thus the error reporting gravely fails. I have posted a fix for this in #9866.

**Comment by straight-shoota** [comment]
I posted a fix in #9867. The doc generator now correctly runs if git is not installed. Can you verify this @myfreeweb? The patch from #9867 should be sufficient, the other PR just fixes a subsequent bug.
